src/system_database.ts

```diff
@@ -946,16 +1105,109 @@ export class PostgresSystemDatabase implements SystemDatabase {
         [workflowID],
       );
 
+      await client.query(
+        `DELETE FROM ${DBOSExecutor.systemDBSchemaName}.workflow_cancel 
+         WHERE workflow_id = $1`,
+        [workflowID],
+      );
+
       // Update status to pending and reset recovery attempts
       await this.recordWorkflowStatusChange(workflowID, StatusString.PENDING, { resetRecoveryAttempts: true }, client);
 
       await client.query('COMMIT');
     } catch (error) {
+      this.logger.error(error);
       await client.query('ROLLBACK');
       throw error;
     } finally {
       client.release();
     }
+    this.clearWFCancelMap(workflowID);
+  }
+
+  // cancel checker
+  private isCancelCheckerRunning: boolean = false;
+  private interruptCancelCheckerResolve?: () => void;
+  private cancelPollerLoopEnded?: Promise<void> = undefined;
+
+  private stopCancelChecker() {
+    if (!this.isCancelCheckerRunning) return;
+    this.isCancelCheckerRunning = false;
+    if (this.interruptCancelCheckerResolve) {
+      this.interruptCancelCheckerResolve();
+    }
+  }
+
+  private startCancelChecker() {
+    this.cancelPollerLoopEnded = this.cancelCheckerLoop();
+  }
+
+  async cancelCheckerLoop(): Promise<void> {
+    this.isCancelCheckerRunning = true;
+    while (this.isCancelCheckerRunning) {
+      // Wait for either the timeout or an interruption
+      let timer: NodeJS.Timeout;
+      const timeoutPromise = new Promise<void>((resolve) => {
+        timer = setTimeout(() => {
+          resolve();
+        }, this.dbPollingIntervalEventMs);
+      });
+      await Promise.race([
+        timeoutPromise,
+        new Promise<void>((_, reject) => (this.interruptCancelCheckerResolve = reject)),
+      ]).catch(() => {
+        this.logger.debug('Cancel check loop interrupted!');
+      }); // Interrupt sleep throws
+      clearTimeout(timer!);
+
+      if (!this.isCancelCheckerRunning) {
+        break;
+      }
+
+      // See which cancels are in the DB
+      const res = await this.pool.query<{ workflow_id: string }>(
+        `SELECT workflow_id FROM ${DBOSExecutor.systemDBSchemaName}.workflow_cancel;`,
```

Comment:
I wonder if we should add a created_at timestamp column in this table and index on that column, and only select workflows that are started recently (after the process start, and then rolling update). We run this loop every 10 seconds, and if someone batch cancels 1 million workflows, we'll repeatedly read that 1 million back every 10 seconds.

Could consider adding it in a separate PR.